internal/util/vault_sa.go

```diff
@@ -0,0 +1,246 @@
+})
+
+// initVaultTenantSA returns an interface to HashiCorp Vault KMS where Tenants
+// use their ServiceAccount to access the service.
+func initVaultTenantSA(args KMSInitializerArgs) (EncryptionKMS, error) {
+	var err error
+
+	config := args.Config
+	if _, ok := config[kmsProviderKey]; ok {
+		// configuration comes from the ConfigMap, needs to be
+		// converted to vaultTokenConf type
+		config, err = transformConfig(config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert configuration: %w", err)
+		}
+	}
+
+	kms := &VaultTenantSA{}
+	err = kms.initConnection(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize Vault connection: %w", err)
+	}
+
+	// set default values for optional config options
+	kms.ConfigName = vaultTokensDefaultConfigName
+	kms.tenantSAName = vaultTenantSAName
+
+	// TODO: should this be configurable per tenant?
+	vaultRole := vaultDefaultRole
+	err = setConfigString(&vaultRole, args.Config, "vaultRole")
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+	kms.vaultConfig[vault.AuthKubernetesRole] = vaultRole
+
+	err = kms.parseConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch the configuration for the tenant
+	if args.Tenant != "" {
+		kms.Tenant = args.Tenant
+		tenantConfig, found := fetchTenantConfig(config, args.Tenant)
+		if found {
+			// override connection details from the tenant
+			err = kms.parseConfig(tenantConfig)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		err = kms.parseTenantConfig()
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse config for tenant: %w", err)
+		}
+
+		err = kms.setServiceAccountName(config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to set the ServiceAccount name from %s for tenant (%s): %w",
+				kms.ConfigName, kms.Tenant, err)
+		}
+	}
+
+	kms.vaultConfig[vault.AuthMethod] = vault.AuthMethodKubernetes
+	kms.vaultConfig[vault.AuthKubernetesTokenPath], err = kms.getTokenPath()
+	if err != nil {
+		return nil, fmt.Errorf("failed setting up token for %s/%s: %w", kms.Tenant, kms.tenantSAName, err)
+	}
+
+	err = kms.initCertificates(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize Vault certificates: %w", err)
+	}
+	// connect to the Vault service
+	err = kms.connectVault()
+	if err != nil {
+		return nil, err
+	}
+
+	return kms, nil
+}
+
+// Destroy removes the temporary stored token from the ServiceAccount and
+// destroys the vaultTenantConnection object.
+func (kms *VaultTenantSA) Destroy() {
+	if kms.saTokenDir != "" {
+		_ = os.RemoveAll(kms.saTokenDir)
+	}
+
+	kms.vaultTenantConnection.Destroy()
+}
+
+// setServiceAccountName stores the name of the ServiceAccount in the
+// configuration if it has been set in the options.
+func (kms *VaultTenantSA) setServiceAccountName(config map[string]interface{}) error {
+	err := setConfigString(&kms.tenantSAName, config, "tenantSAName")
+	if errors.Is(err, errConfigOptionInvalid) {
```

Comment:
can we also log this error ?

Reply:
it is logged, the calling functions will take care of that